resources/build/test/testing-framework.inc.sh

```diff
@@ -0,0 +1,261 @@
+
+  if [[ "${actual}" != "${expected}" ]]; then
+    messages+=("- Actual ${variable}should equal expected:
+                             Actual  : '${actual}'
+                             Expected: '${expected}'
+
+                          ")
+    ((test_result++))
+  fi
+}
+
+assert-contains() {
+  local actual="$1"
+  local expected="$2"
+  local variable=
+  if [[ $# -gt 2 ]]; then
+    variable="$3 "
+  fi
+
+  if [[ "${actual}" != *"${expected}"* ]]; then
+    messages+=("- Actual ${variable}should contain expected:
+                             Actual  : '${actual}'
+                             Expected: '${expected}'
+
+                          ")
+    ((test_result++))
+  fi
+}
+
+assert-not-contains() {
+  local actual="$1"
+  local expected="$2"
+  local variable=
+  if [[ $# -gt 2 ]]; then
+    variable="$3 "
+  fi
+
+  if [[ "${actual}" == *"${expected}"* ]]; then
+    messages+=("- Actual ${variable}should not contain expected:
+                             Actual  : '${actual}'
+                             Expected: '${expected}'
+
+                          ")
+    ((test_result++))
+  fi
+}
+
+assert-failed() {
+  local status="$1"
+
+  if [[ "${status}" == 0 ]]; then
+    messages+=("- Expected test to fail but it succeeded
+                          ")
+    ((test_result++))
+  fi
+}
+
+assert-succeeded() {
+  local status="$1"
+
+  if [[ "${status}" != 0 ]]; then
+    messages+=("- Expected test to succeed but it failed
+                          ")
+    ((test_result++))
+  fi
+}
+
+# setup() will run before each test
+# can be overriden in test script
+setup() {
+  true
+}
+
+# teardown() will run after each test
+# can be overriden in test script
+teardown() {
+  true
+}
+
+# setup_file() will run once before all tests in the file
+# can be overriden in test script
+setup_file() {
+  true
+}
+
+# teardown_file() will run once after all tests in the file
+# can be overriden in test script
+teardown_file() {
+  true
+}
+
+# Discover and run all tests in the current script
+#
+# Parameters:
+#  [$1 $2] / [first last] - optional range of lines in the tests script to search for tests.
+#                           This helps to split a test file into several sections where
+#                           each section can have separate setup/teardown functions.
+#
+# 'run_tests' looks for tests that follow the following pattern:
+#   function test_name() {
+#     test code
+#   }
+run_tests() {
+  local test_count test_fail_count line func
+  local first="${1:-1}"
+  local last="${2:-$(wc -l < "${BASH_ARGV0}")}"
+
+  test_count=0
+  test_fail_count=0
+
+  setup_file
+
+  # shellcheck disable=SC2312
+  while read -r line; do
+    func="$(_get_test_func "${line}")"
+    _run_single_test "${func}"
+  done < <(sed -n "${first},${last}p" "${BASH_ARGV0}" | grep -e "^\s*function\s*test_")
+
+  teardown_file
+
+  builder_echo green "Total of ${test_count} tests run, $((test_count - test_fail_count)) passed, ${test_fail_count} failed"
```

Comment:
Does the `run_tests()` function return a failure code if one or more tests fail?  If so, I'm having trouble seeing where.

Reply:
Good point. Fixed. Thanks!